### PR TITLE
Introduce proper lifecycle-handling. Rework communication stack.

### PR DESCRIPTION
**Pull request by FinnStutzenstein**

### CODE REVIEW

File: client/src/app/app.component.ts [hunk 1]
```diff
@@ -88,111 +75,35 @@ export class AppComponent {
...
                 filter(s => s),
                 take(1)
             )
-            .subscribe(() => servertimeService.startScheduler());
+            .toPromise();
+        await this.onInitDone;
+
+        setTimeout(() => {
+            this.lifecycleService.appLoaded.next();
+            /*this.openslidesService.bootup();
```
Comment (FinnStutzenstein) [inline client/src/app/app.component.ts:97]: TODO: remove

File: client/src/app/app.component.ts [hunk 2]
```diff
@@ -88,111 +75,35 @@ export class AppComponent {
...
-     */
-    private overloadModulo(): void {
-        Object.defineProperty(Number.prototype, 'modulo', {
-            value: function (n: number): number {
-                return ((this % n) + n) % n;
-            },
-            enumerable: false
-        });
+    public ngOnInit(): void {
+        /*setTimeout(() => {
```
Comment (FinnStutzenstein) [inline client/src/app/app.component.ts:103]: TODO: remove

File: client/src/app/core/core-services/lifecycle.service.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { EventEmitter, Injectable } from '@angular/core';
+
+@Injectable({
+    providedIn: 'root'
+})
+export class LifecycleService {
```
Comment (FinnStutzenstein) [inline client/src/app/core/core-services/lifecycle.service.ts:6]: TODO: Docs

**Comment by FinnStutzenstein** [comment]
@jsangmeister Works best with https://github.com/OpenSlides/OpenSlides/pull/5487. You can start/stop the autoupdateservice with `./dc-dev.sh start autoupdate`/`./dc-dev.sh stop autoupdate` to simulate a down server.

File: client/src/app/core/core-services/openslides.service.ts [hunk 4]
```diff
@@ -88,84 +87,39 @@ export class OpenSlidesService {
...
-        await this.bootup();
+    public reboot(): void {
+        this.shutdown();
+        this.bootup();
     }
 
     /**
      * Verify that the operator is the same as it was before. Should be alled on a reconnect.
      *
      * @returns true, if the user is still logged in
```
Comment (jsangmeister) [inline client/src/app/core/core-services/openslides.service.ts:115]: remove docs if no longer correct

File: client/src/app/core/core-services/streaming-communication.service.ts [hunk 5]
```diff
@@ -1,209 +1,158 @@
...
+            delete this.openStreams[streamContainer.id];
+            this.goOffline(streamContainer.endpoint);
+        } else {
+            // retry it after some time:
+            try {
+                if (await this.delayAndCheckReconnection()) {
+                    await this.connect(streamContainer);
+                }
+            } catch (e) {
+                // connectWithWrapper can throw an OfflineError. This one can be ignored.
```
Comment (jsangmeister) [inline client/src/app/core/core-services/streaming-communication.service.ts:82]: Update comment

File: client/src/app/core/core-services/streaming-communication.service.ts [hunk 6]
```diff
@@ -1,209 +1,158 @@
...
+        }
 
-    private handleMessage(event: HttpDownloadProgressEvent): void {
-        if (this.hasError) {
-            if (!this.gotFirstResponse.wasResolved) {
-                this._errorContent = event.partialText;
-                this.gotFirstResponse.resolve(this.hasError);
-            }
-            return;
+        // do not continue, if we are offline!
```
Comment (jsangmeister) [inline client/src/app/core/core-services/streaming-communication.service.ts:102]: C&P? ;)

File: client/src/app/core/core-services/streaming-communication.service.ts [hunk 7]
```diff
@@ -1,209 +1,158 @@
...
+        }
+    }
 
     /**
-     * This index holds always the position of the current JOSN-packet, that we are receiving.
+     * Returns true, if a reconnect attempt should be done.
      */
-    private contentStartIndex = 0;
+    private async delayAndCheckReconnection(): Promise<boolean> {
+        const delay = Math.floor(Math.random() * 3000 + 2000);
```
Comment (jsangmeister) [inline client/src/app/core/core-services/streaming-communication.service.ts:91]: Is there a reason to randomize the delay instead of a constant?

Comment (FinnStutzenstein) [reply client/src/app/core/core-services/streaming-communication.service.ts:91]: This was introduced to not instantly DDOS our servers on a mass reconnection. We tried to spread the requests a bit.

**Pull request merged**